client/src/components/BarChart.jsx

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import {BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, Legend} from 'recharts';
+import { FormGroup, FormControl, InputGroup, Glyphicon } from 'react-bootstrap';
+
+class BarChartComponent extends React.Component {
+	constructor(props){
+		super(props);
+		this.state = {
+			organizationCriteria: 'alphabetical',
+			showResults: false,
+			ordering: 'descending',
+			chartData: []
+		}
+
+		this.search = this.search.bind(this);
+		this.updateSearchValue = this.updateSearchValue.bind(this);
+		this.frequencySort = this.frequencySort.bind(this);
+	}
+
+	componentDidMount() {
+		fetch('/api/chartingData')
+		.then(response => response.json())
+		.then(body => {
+			const data = body.data.map(entry => {
+				return {name: entry.displayables[0].displayInfo.name, 'total listens': entry.value}
+			});
+			this.setState({chartData: data});
+		})
+		.catch(error => {
+			console.log(error);
+			this.setState({error: true});
+		});
+	}
+
+	alphabeticalSort(a, b) {
+		return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1;
+	}
+	
+	frequencySort(a, b) {
+		if (this.state.ordering === 'descending') {
+			return a['total listens'] > b['total listens'] ? -1 : 1;
+		}
+		if (this.state.ordering === 'ascending') {
+			return a['total listens'] > b['total listens'] ? 1 : -1;
+		}
+	}
+
+	organizeBy(criteria) {
+		let sortedData;
+		if (criteria === 'frequencyOfListen') {
+			sortedData = this.state.chartData.slice().sort(this.frequencySort);
+		}
+
+		if (criteria === 'alphabetical') {
+			sortedData = this.state.chartData.slice().sort(this.alphabeticalSort);
+		}
+
+		this.setState({sortedChartData: sortedData, ordering: this.state.ordering === 'ascending' ? 'descending' : 'ascending'})
+	}
+
+	renderErrorMessage() {
+		return (
+			<span class="errorNotice">Error retrieving data from server. Please try again later.</span>
+		)
+	}
+
+
+	search() {
+
+		const query = this.state.query;
+		this.state.chartData.find(record => {
+			if (record.name.toLowerCase() === query.toLowerCase()) {
+				this.setState({showResults: true, query: record.name, queryResults: record['total listens']}); 
+			} 
+		})
+	}
+
+	showResults() {
+		return (
+			<div class="searchResults"> You've listened to {this.state.query} {this.state.queryResults} times </div> 
```

Comment:
I've made a small change which will notify the user, in the same way it would if records were found, if the query string they provide doesn't match any of the existing data records. There's still a small problem when a user submits an empty string (it doesn't break, but it treats the empty string as the query and the notification makes little sense). One possible option here would be to incorporate the modal component I used in the Line Chart component to notify users that the input string was invalid. Alternatively, I could do some more conditional rendering, but it starts to get messy when there are too many nested ternary statements.